Add Soviet-08b

Smittytron opened the pull request:
Most of the code is equivalent to Soviet-08a. This map is much easier than 08a as it has ore mines, so I've upped the attack rates to compensate.

Edit: Added a commit to Update Soviet-08a with any improvements found here.

Smittytron committed:
Add Soviet-08b

Smittytron committed:
Update Soviet-08a

Smittytron committed:
Remove worthless function from Allies08

Smittytron commented:
I've added a commit to remove a worthless function from Allies 08. I can split that off if needed but this is how these kind of fixes were done in the d2k mission PRs.

cjshmyr reviewed (commented):
Completed it on normal without trouble. Felt true to the original, I enjoyed it!

Feedback:
- Enemy harvesters sat in their base doing nothing, intended?
- Can we move the player's sub pen a bit? A purchased sub got stuck here.

  ![image](https://user-images.githubusercontent.com/566742/84714823-12f9bd00-af2c-11ea-805f-bc6340e474d9.png)

Potential polish (doesn't stop an approval from me)
- Spent 6 of 28 minutes hunting down ships at the end after I owned the map, lingering gunboats mostly.

I'll try it on hard mode later. I also need to install movies again before a final review.

pchote commented:
> Spent 6 of 28 minutes hunting down ships at the end after I owned the map, lingering gunboats mostly.

I suspect that the original mission would have avoided this by having the script throw everything they have left at the player once beaten hard enough. We could/should do the same here?

cjshmyr commented:
The original soviet 8b didn't unfortunately. Another idea is to send a couple of idling boats at the subpen area periodically.

Smittytron commented:
Updated: Subpen moved and Harvester scan range increased.

> I suspect that the original mission would have avoided this by having the script throw everything they have left at the player once beaten hard enough. We could/should do the same here?

Boats don't seem to want to `Hunt()`. I could have them attackmove the subpen area on the destruction of the enemy base if that seems acceptable. (I'm envisioning an odd situation where the player didn't use subs at all and ten boats stack up at the player's doorstep at the end of the game.)
Could also go with hamb's idea and have a boat attack towards the subpen every few minutes.

cjshmyr commented:
I think that odd situation will happen most playthroughs. Asked if pchote had a preference:
> \<pchote> either is fine IMO, maybe tie it to difficulty levels?
> \<hamb> pchote: periodic with shorter intervals if harder?
> \<pchote> yeah

Let's go for periodic and could scale interval by difficulty

Smittytron commented:
Rebased and updated to work with #18271.

I've added periodic gunboat attacks every 2-3 minutes. There are only six boats in 08b and seven in 08a so I'm going to forgo tying them to difficulty.

abcdefg30 commented on mods/ra/maps/allies-08a/allies08a.lua:
```diff
@@ -74,18 +74,6 @@ CreateScientists = function()
 	end)
 end
 
-FinishTimer = function()
```
Why are we removing this instead of using it?

Smittytron replied:
Since the mission ends within seconds of the timer finishing, there is no need for the text to alert the player by flashing or for the text to disappear later. I honestly didn't know how this function worked when I copy/pasted it a couple years ago. 🤦

abcdefg30 replied:
Ok, that explains why we remove it. ^^

Smittytron committed:
Move Harvester overrides to campaign-rules.yaml

Smittytron commented:
We've run into a couple more missions where harvesters just sit there (likely from #16920) so I'm moving the harvester overrides to campaign-rules.yaml in a new commit. A search range of 50 seems overkill but long-distance mining and maps without oremines are a common feature of missions.

cjshmyr reviewed (dismissed):
Finished on hard, nice work!

abcdefg30 commented on mods/ra/maps/soviet-08b/soviet08b-AI.lua:
```diff
@@ -0,0 +1,176 @@
+			patrols.Patrol(DDPatrol2Path, true, 200)
+		end)
+	end)
+	if GreeceNavalYard.IsDead then
```
Shouldn't that be checked inside the trigger? I assume the idea here was to not have additional reinforcements once the naval yard is destroyed.

abcdefg30 commented on mods/ra/maps/soviet-08b/soviet08b.lua:
```diff
@@ -0,0 +1,176 @@
+   the License, or (at your option) any later version. For more
+   information, see COPYING.
+]]
+alliedScouts = { Jeep1, Jeep2, Rifle1, Rifle2, Rifle3, Rifle4 }
```
Style nit: Global variables should start with a capital letter. (Applies to several in this file.)

abcdefg30 commented on mods/ra/maps/soviet-08b/soviet08b.lua:
```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+Obj2ActorTriggerActivator = { Church, Civ1, Civ2, Civ3, Civ4, Civ5, Civ6, Civ7, Civ8 }
```
Can we give this a more descriptive name?

abcdefg30 commented on mods/ra/maps/soviet-08b/soviet08b.lua:
```diff
@@ -0,0 +1,176 @@
+	Actor.Create("moneycrate", true, { Owner = ussr, Location = ChurchAmbush.Location })
+end)
+
+Obj2TriggerFunction = function()
```
Same here.

abcdefg30 commented on mods/ra/maps/soviet-08b/soviet08b.lua:
```diff
@@ -0,0 +1,176 @@
+end
+
+Trigger.OnKilled(Church, function()
+	Actor.Create("moneycrate", true, { Owner = ussr, Location = ChurchAmbush.Location })
```
I forgot if we disabled civilians spawning from the church already. Did we?

Smittytron replied:
We haven't. So far I haven't observed a civ picking up this crate. I think that has to do with creating the actor here instead of spawning one on death with a rules override.

abcdefg30 replied:
I guess you were just lucky since I don't see the spawners disabled here. Can you apply a fix like #18318?

Smittytron replied:
Done

Smittytron commented:
Updated

abcdefg30 commented on mods/ra/maps/soviet-08b/soviet08b.lua:
```diff
@@ -0,0 +1,176 @@
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "New " .. string.lower(p.GetObjectiveType(id)) .. " objective")
+	end)
+	
+	KillAll = ussr.AddPrimaryObjective("Destroy all Allied units and structures.")
```
This should use the new `AddObjective` api.

abcdefg30 commented on mods/ra/maps/soviet-08b/soviet08b.lua:
```diff
@@ -0,0 +1,176 @@
+ChurchAmbushTrigger = function()
+	if not AmbushSwitch then
+		local hiding = Reinforcements.Reinforce(germany, { 'e1', 'e1', 'e1', 'e1', 'e1', 'e1', 'e1', 'e3', 'e3', 'e3' }, { ChurchAmbush.Location, AmbushMove.Location }, 0)
+		Utils.Do(hiding, function(actor)
```
Since `IdleHunt` is a function that takes one parameter of type actor we can simplify this to `Utils.Do(hiding, IdleHunt)`.

Smittytron commented:
Updated

Smittytron commented:
Updated for case conventions from other mission reviews

cjshmyr commented on mods/ra/maps/soviet-08b/soviet08b-AI.lua:
```diff
@@ -0,0 +1,178 @@
+
+WTransDelays =
+{
+	easy = 4,
```
Would it be consistent to wrap these with `DateTime.Minutes()` like the intervals / delays below?

cjshmyr commented on mods/ra/maps/soviet-08b/soviet08b-AI.lua:
```diff
@@ -0,0 +1,178 @@
+	end)
+end
+
+BringDDPatrol1 = function()
```
Identical to `BringDDPatrol2` except for the patrol actors and patrol path. It would be possible to make those parameters for this function to remove duplication.

If this one is too much trouble I'm ok with it as-is personally, since it exists already in soviet-08a

Smittytron replied:
I'll need to do some learning to figure out how to accomplish this.

cjshmyr replied:
I have an example of it @ https://github.com/cjshmyr/OpenRA/commit/834debf2312c8f88788fe6a4d0c098c31e2ba95a

cjshmyr commented on mods/ra/maps/soviet-08b/soviet08b-AI.lua:
```diff
@@ -0,0 +1,178 @@
+
+	Trigger.AfterDelay(DateTime.Minutes(3), WTransWaves)
+
+	Trigger.AfterDelay(AlliedVehiclesUpgradeDelay, function() AlliedVehicleType = "Upgraded" end)	
```
The above delay isn't declared globally but this one is -- what is the preference on this?

abcdefg30 replied:
Often things that are to be changed by difficulties are declared global.

cjshmyr commented on mods/ra/maps/soviet-08b/soviet08b.lua:
```diff
@@ -0,0 +1,174 @@
+	Trigger.OnEnteredProximityTrigger(Jeep2.CenterPosition, WDist.FromCells(12), function(actor, id)
+		if actor.Owner == USSR and actor.Type == "barr" then
+			AlliedScouts = Utils.Where(AlliedScouts, function(scout) return not scout.IsDead end)
+			local removed
```
Assuming our convention is `local removed = false`?

abcdefg30 replied:
I'd say either is ok, we use the current one in a few places iirc.

cjshmyr commented on mods/ra/maps/soviet-08b/soviet08b.lua:
```diff
@@ -0,0 +1,174 @@
+	Trigger.AfterDelay(DateTime.Minutes(14), function() BoatAttack(Gunboat6) end)
+end
+
+OnAnyDamaged = function(actors, func)
```
Should we add this to the Lua API `Trigger.OnAnyDamaged`?

Smittytron replied:
That or in campaign-global.lua
Either way I'll be using it in future missions.

cjshmyr reviewed (commented):
Played through again on normal 👍

cjshmyr reviewed (dismissed):
👍

Smittytron commented:
Thanks, updated.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f401d5286c6309c287efa6d0c28394215731c9c6...f49b92785c2eae1b4a1cf08453b25741abb88e47)
